rskj-core/src/main/java/co/rsk/net/AsyncNodeBlockProcessor.java

```diff
@@ -0,0 +1,181 @@
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * GNU Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program. If not, see <http://www.gnu.org/licenses/>.
+ */
+
+package co.rsk.net;
+
+import co.rsk.config.InternalService;
+import co.rsk.crypto.Keccak256;
+import co.rsk.net.sync.SyncConfiguration;
+import co.rsk.validators.BlockValidator;
+import org.ethereum.core.Block;
+import org.ethereum.core.Blockchain;
+import org.ethereum.core.ImportResult;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import javax.annotation.Nonnull;
+import javax.annotation.Nullable;
+import java.time.Duration;
+import java.time.Instant;
+import java.util.Collections;
+import java.util.Map;
+import java.util.concurrent.BlockingQueue;
+import java.util.concurrent.LinkedBlockingQueue;
+
+/**
+ * AsyncNodeBlockProcessor processes blockchain blocks that are received from other nodes.
+ * If a block passes validation, it will immediately be propagated to other nodes and its execution will be scheduled on a separate thread.
+ * Blocks are being executed and connected to the blockchain sequentially one after another.
+ * <p>
+ * If a block is not ready to be added to the blockchain, it will be on hold in a BlockStore.
+ * <p>
+ */
+public class AsyncNodeBlockProcessor extends NodeBlockProcessor implements InternalService, Runnable {
+
+    private static final Logger logger = LoggerFactory.getLogger("asyncblockprocessor");
+
+    private final BlockingQueue<PeerBlockPair> blocksToProcess = new LinkedBlockingQueue<>();
+
+    private final BlockValidator blockRelayValidator;
+
+    private final Thread thread = new Thread(this,"async block processor");
+
+    private final Listener listener;
+
+    private volatile boolean stopped;
+
+    public AsyncNodeBlockProcessor(@Nonnull NetBlockStore store, @Nonnull Blockchain blockchain, @Nonnull BlockNodeInformation nodeInformation,
+                                   @Nonnull BlockSyncService blockSyncService, @Nonnull SyncConfiguration syncConfiguration,
+                                   @Nonnull BlockValidator blockRelayValidator, @Nullable Listener listener) {
+        super(store, blockchain, nodeInformation, blockSyncService, syncConfiguration);
+        this.listener = listener;
+        this.blockRelayValidator = blockRelayValidator;
+    }
+
+    @Override
+    public BlockProcessResult processBlock(@Nullable Peer sender, @Nonnull Block block) {
+        final Instant start = Instant.now();
+
+        boolean looksGood = blockSyncService.preprocessBlock(block, sender, false);
```

Comment:
I've just inverted the order, as pato mentions, and I didn't got any error, that's because of `DummyValidator`. We should take care of that

```java
if (isValid(block)) {
            boolean looksGood = blockSyncService.preprocessBlock(block, sender, false);
            if (looksGood) {
```

Reply:
we should decouple that logic (maybe have a BlockBasicValidator which validates genesis, header and block). For me it doesn't make sense to preprocess a block with invalid header

Reply:
I didn't follow all the discussion, but I think we should do all the validations that NOT involves access to disk or memory caches, FIRST.